src/Datadog.Trace/Agent/AgentWriter.cs

```diff
@@ -62,57 +118,295 @@ public async Task FlushAndCloseAsync()
+            if (!buffer.Lock())
+            {
+                // Buffer is already locked, it's probably being flushed from another thread
+                return;
+            }
+
+            try
+            {
+                if (_statsd != null)
+                {
+                    _statsd.Increment(TracerMetricNames.Queue.DequeuedTraces, buffer.TraceCount);
+                    _statsd.Increment(TracerMetricNames.Queue.DequeuedSpans, buffer.SpanCount);
+                }
+
+                if (buffer.TraceCount > 0)
+                {
+                    await _api.SendTracesAsync(buffer.Data, buffer.TraceCount).ConfigureAwait(false);
+                }
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, "An unhandled error occurred while flushing a buffer");
+            }
+            finally
+            {
+                // Clear and unlock the buffer
+                buffer.Clear();
+            }
+        }
+
+        private void SerializeTrace(Span[] trace)
+        {
+            // Declaring as inline method because only safe to invoke in the context of SerializeTrace
+            SpanBuffer SwapBuffers()
+            {
+                if (_activeBuffer == _frontBuffer)
+                {
+                    if (!_backBuffer.IsFull)
                     {
-                        await FlushTracesAsync().ConfigureAwait(false);
-                        return;
+                        Volatile.Write(ref _activeBuffer, _backBuffer);
+                        return _activeBuffer;
                     }
-                    else
+                }
+                else
+                {
+                    if (!_frontBuffer.IsFull)
                     {
-                        await FlushTracesAsync().ConfigureAwait(false);
+                        Volatile.Write(ref _activeBuffer, _frontBuffer);
+                        return _activeBuffer;
+                    }
+                }
+
+                return null;
+            }
+
+            // We use a double-buffering mechanism
+            // This allows the serialization thread to keep doing its job while a buffer is being flushed
+            var buffer = _activeBuffer;
+
+            if (buffer.TryWrite(trace, ref _temporaryBuffer))
+            {
+                // Serialization to the primary buffer succeeded
+                return;
+            }
+
+            // Active buffer is full, swap them
+            buffer = SwapBuffers();
+
+            if (buffer != null)
+            {
+                // One buffer is full, request an eager flush
+                RequestFlush();
+
+                if (buffer.TryWrite(trace, ref _temporaryBuffer))
+                {
+                    // Serialization to the secondary buffer succeeded
+                    return;
+                }
+            }
+
+            // All the buffers are full :( drop the trace
+            if (_statsd != null)
```

Comment:
```suggestion
            Log.Warning("Trace buffer is full. Dropping a trace from the buffer.");
            if (_statsd != null)
```

Reply:
@zacharycmontoya: this message sounds better with the current implementation, where we drop a random trace that is already buffered (`"...from the buffer"`). Instead, we now stop buffering any new traces. Maybe we can tweak the message to something like:
- Trace buffer is full. Dropping trace.
- Trace buffer is full. Rejecting trace.
- Trace buffer is full. Cannot buffer more traces.

Reply:
Done in https://github.com/DataDog/dd-trace-dotnet/pull/1151/commits/036ae23d5ace2c8e19264ac624a2c0ec63de27e6